src/rust/engine/process_execution/src/remote_cache.rs

```diff
@@ -0,0 +1,451 @@
+/// has a hit. On the way back out of the stack, the result will be stored remotely and
+/// then locally.
+#[derive(Clone)]
+pub struct CommandRunner {
+  underlying: Arc<dyn crate::CommandRunner>,
+  metadata: ProcessMetadata,
+  store: Store,
+  action_cache_client: Arc<bazel_protos::remote_execution_grpc::ActionCacheClient>,
+  headers: BTreeMap<String, String>,
+  platform: Platform,
+  cache_read: bool,
+  cache_write: bool,
+}
+
+impl CommandRunner {
+  pub fn new(
+    underlying: Arc<dyn crate::CommandRunner>,
+    metadata: ProcessMetadata,
+    store: Store,
+    action_cache_address: &str,
+    root_ca_certs: Option<Vec<u8>>,
+    oauth_bearer_token: Option<String>,
+    headers: BTreeMap<String, String>,
+    platform: Platform,
+    cache_read: bool,
+    cache_write: bool,
+  ) -> Result<Self, String> {
+    let env = Arc::new(grpcio::EnvBuilder::new().build());
+    let channel = {
+      let builder = grpcio::ChannelBuilder::new(env);
+      if let Some(ref root_ca_certs) = root_ca_certs {
+        let creds = grpcio::ChannelCredentialsBuilder::new()
+          .root_cert(root_ca_certs.clone())
+          .build();
+        builder.secure_connect(action_cache_address, creds)
+      } else {
+        builder.connect(action_cache_address)
+      }
+    };
+    let action_cache_client = Arc::new(
+      bazel_protos::remote_execution_grpc::ActionCacheClient::new(channel),
+    );
+
+    let mut headers = headers;
+    if let Some(oauth_bearer_token) = oauth_bearer_token {
+      headers.insert(
+        String::from("authorization"),
+        format!("Bearer {}", oauth_bearer_token.trim()),
+      );
+    }
+
+    // Validate any configured static headers.
+    call_option(&headers, None)?;
+
+    Ok(CommandRunner {
+      underlying,
+      metadata,
+      store,
+      action_cache_client,
+      headers,
+      platform,
+      cache_read,
+      cache_write,
+    })
+  }
+
+  /// Create a REAPI `Tree` protobuf for an output directory by traversing down from a Pants
+  /// merged final output directory to find the specific path to extract. (REAPI requires
+  /// output directories to be stored as `Tree` protos that contain all of the `Directory`
+  /// protos that constitute the directory tree.)
+  async fn make_tree_for_output_directory(
+    &self,
+    root_directory_digest: Digest,
+    directory_path: RelativePath,
+    store: &Store,
+  ) -> Result<Tree, String> {
+    // Traverse down from the root directory digest to find the directory digest for
+    // the output directory.
+    let mut current_directory_digest = root_directory_digest;
+    for next_path_component in directory_path.as_ref().components() {
+      let next_name = match next_path_component {
+        Component::Normal(name) => name
+          .to_str()
+          .ok_or_else(|| format!("unable to convert '{:?}' to string", name))?,
+        _ => return Err("illegal state: unexpected path component in relative path".into()),
```

Comment:
nit: it'd be good to capitalize this + the other error messages here.

Reply:
Done.